Fixes encoded values on Blazor Router

javiercn opened the pull request:
Fix handling of encoded `/` in URLs within the Blazor router.

Fixes https://github.com/dotnet/aspnetcore/issues/52808, https://github.com/dotnet/aspnetcore/issues/53138, https://github.com/dotnet/aspnetcore/issues/53262

In .NET 8.0 we unified the routing system for ASP.NET Core and Blazor in order to bridge the functionality gaps and provide a consistent experience for SSR and interactive routing. 
As part of such, we had to make adjustments to account for some of the quirks of the Blazor router (where it would behave slightly different than the ASP.NET Core router).
As part of this, we were decoding the entire URL before passing it to the Blazor router instead of (as it did before), splitting each segment and decoding each segment individually.

This introduced a regression when using URLs that contained encoded segments with `/` in them, as the router would then treat the `/` as a path separator instead of part of the segment and we would not be able to route correctly.

To fix this, we now decode each segment individually inside the router when the code runs as part of Blazor. In addition to that, when the routing happens server side, we now post-process the resulting values to URL decode them so that they match the behavior of the Blazor router.

To summarize, here is the before and after behavior:

Before:
Framework version and router | pattern | request | result
|--- | --- | --- | --- |
7.0 interactive | `/some/{parameter}` | `/some/encoded%2Fvalue` | `parameter = "encoded/value"`
8.0 SSR | `/some/{parameter}` | `/some/encoded%2Fvalue` | `parameter = "encoded%2Fvalue"`
8.0 interactive | `/some/{parameter}` | `/some/encoded%2Fvalue` | Not Found (URI got decoded as /some/encoded/value)

After:
Framework version and router | pattern | request | result
|--- | --- | --- | --- |
7.0 interactive | `/some/{parameter}` | `/some/encoded%2Fvalue` | `parameter = "encoded/value"`
8.0 SSR | `/some/{parameter}` | `/some/encoded%2Fvalue` | `parameter = "encoded/value"`
8.0 interactive | `/some/{parameter}` | `/some/encoded%2Fvalue` | `parameter = "encoded/value"`

SteveSandersonMS commented on src/Components/Components/src/Routing/RouteTableFactory.cs:
```diff
@@ -112,9 +113,11 @@ private static string[] GetTemplates(Type componentType)
     internal static RouteTable Create(Dictionary<Type, string[]> templatesByHandler, IServiceProvider serviceProvider)
     {
+        var routeOptions = Options.Create(new RouteOptions());
+        routeOptions.Value.SetParameterPolicy("regex", typeof(RegexInlineRouteConstraint));
```
I'm unsure if this is part of dealing with encoding or if it's addressing a separate issue. Could you clarify?

javiercn replied:
This is addressing a separate issue, we weren't wiring up the regex, which was needed to test this fully, as it doesn't affect other constraints that the Blazor router supports and we don't have support for authoring custom constraints either.

SteveSandersonMS replied:
Have we effectively added a whole new feature with this line of code then? I'm totally fine with this being added, assuming it doesn't appreciably impact the default bundle size for a Blazor WebAssembly app.

Could you confirm if that's the case?

javiercn replied:
I'm in the process of confirming the bundle size, it's not a feature, it's a bug fix IMO. It works on SSR routing but won't work once it becomes interactive.

SteveSandersonMS replied:
Thanks, please do confirm! I think we still have a choice of whether to include that feature for interactive mode, given that it's never worked on interactive mode in the past.

javiercn replied:
It turns out that this regressed the size, so I'm going to take it out into its own separate PR that enables this conditionally on WebAssembly behind a feature flag. The feature flag will also need a change in the Blazor SDK to control it, but for the purposes of this repo, we'll keep it enabled.

![image](https://github.com/dotnet/aspnetcore/assets/6995051/03d150a3-6edb-4873-94db-06cdd474a793)

SteveSandersonMS commented on src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/Routing/Encoded/EncodedCatchAll.razor:
```diff
@@ -0,0 +1,25 @@
+﻿@page "/🙂/routing/catch-all/{*parameter}"
```
If I ever see a URL like this in the wild, I'll know it's time to go live off-grid and pretend never to have heard of computers...

SteveSandersonMS commented on src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/Routing/Encoded/EncodedConstrainedCatchAll.razor:
```diff
@@ -0,0 +1,24 @@
+﻿@page "/🙂/routing/constrained-catch-all/{*parameter:regex(http:%2F%2Fwww\\.example\\.com%2Flogin%2Fcallback\\/another)}"
```
![My eyes](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYmFlcWM2aTRyMmowejJlaTZjYTN1empjOXh5YnU2OTBxOXc5YjVsNSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/oS4bJ2sRXKVDa/giphy.gif)

javiercn replied:
Yeah, this is a bit extreme, but I wanted to make sure we cover all the edge cases correctly, as this is very tricky.

SteveSandersonMS reviewed (approved):
Great!

javiercn committed:
[Blazor] Fixed encoded values on Blazor Router

As part of the routing unification process we switched the way we were
decoding the URL prior to feeding it to routing and that introduced a
small regression in interactive routing compared to .NET 7.0.

This commit fixes that regression by using the same logic for decoding
the URL in the client that is used on the server.

In addition to that, the Blazor router now post processes the URL to
replace instances of `%2F` with `/` when providing values to maintain
the behavior in 7.0 where it used UnescapeDataString.

This also makes the routing on the server and on the client consistent
in their handling of encoded `/` characters.

javiercn commented:
/backport to release/8.0
